File: Python/2buy.py
"""
This Code Will Buy 2 ADA or Any coin you want 
One line 82-85 You can Change a Coin and Buy amount  
For More information read a API.md
"""


import requests
import hmac
import hashlib
from urllib.parse import urlencode

API_KEY = input("Please Enter You API Key : ")
SECRET_KEY = input("Please Enter You Secret Key : ")
BASE_URL = "https://rest.bit24.cash"

def sign_params(params: dict) -> str:
    """
    Create HMAC‑SHA256 signature for the given parameters.
    Parameters are sorted alphabetically, joined with '&', then signed.
    """
    params_to_sign = {k: v for k, v in params.items() if k != "signature"}
    sorted_params = sorted(params_to_sign.items())
    query_string = "&".join(f"{k}={v}" for k, v in sorted_params)
    signature = hmac.new(
        SECRET_KEY.encode('utf-8'),
        query_string.encode('utf-8'),
        hashlib.sha256
    ).hexdigest()
    return signature

def get_best_ask(symbol_base: str, symbol_quote: str) -> str:
    """
    Return the best ask price as a string (exactly as returned by the API).
    """
    url = f"{BASE_URL}/pro/capi/v1/markets/order-books"
    headers = {
        "Accept": "application/json",
        "X-BIT24-APIKEY": API_KEY,
    }
    params = {"base_coin": symbol_base, "quote_coin": symbol_quote}
    resp = requests.get(url, headers=headers, params=params)
    resp.raise_for_status()
    data = resp.json()
    if not data.get("success"):
        raise Exception(f"Order book error: {data.get('error')}")

    sell_orders = data["data"]["sell_orders"]
    if not sell_orders:
        raise Exception("No sell orders available.")
    # Return the price string directly – no float conversion!
    return sell_orders[0]["price"]

def submit_limit_buy_order(symbol_base: str, symbol_quote: str, amount: str, price: str):
    """
    Place a limit buy order. amount and price are strings.
    """
    url = f"{BASE_URL}/pro/capi/v1/orders/submit"
    headers = {
        "Accept": "application/json",
        "Content-Type": "application/x-www-form-urlencoded",
        "X-BIT24-APIKEY": API_KEY,
    }

    params = {
        "base_coin_symbol": symbol_base,
        "quote_coin_symbol": symbol_quote,
        "type": "1",               # buy
        "category_type": "0",      # limit
        "price": price,
        "amount": amount,
    }

    params["signature"] = sign_params(params)

    resp = requests.post(url, headers=headers, data=params)
    # If not OK, print the response to see the exact validation error
    if not resp.ok:
        print("Server responded with error:")
        print(resp.text)
    resp.raise_for_status()
    return resp.json()

def buy_2_ada_on_irt():
    base = "ADA"
    quote = "IRT"
    amount = "2"            # send as simple integer string

    print("Fetching best ask for ADA/IRT ...")
    best_ask_str = get_best_ask(base, quote)
    print(f"Best ask price (raw): {best_ask_str} IRT")

    print(f"Placing limit buy order for {amount} ADA at {best_ask_str} IRT ...")
    response = submit_limit_buy_order(base, quote, amount, best_ask_str)

    if response.get("success"):
        order = response["data"]["order"]
        print(f"Order placed successfully! Order ID: {order['id']}")
        print(f"Status: {order['status_text']}")
        print(f"Price: {order['each_price']} IRT, Amount: {order['amount']} ADA")
        print(f"Total: {order['total']} IRT")
    else:
        print(f"Order failed: {response.get('error')}")

if __name__ == "__main__":
    try:
        buy_2_ada_on_irt()
    except Exception as e:
        print(f"An error occurred: {e}")

File: Go/2buy.go
// This Code Will Buy 2 ADA or Any coin you want
// Change base, quote and amount inside buy2ADAOnIRT()
//
// Run:
// go run main.go

package main

import (
	"bufio"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"sort"
	"strings"
)

const BASE_URL = "https://rest.bit24.cash"

var (
	API_KEY    string
	SECRET_KEY string
)

type OrderBookResponse struct {
	Success bool `json:"success"`
	Data    struct {
		SellOrders []struct {
			Price string `json:"price"`
		} `json:"sell_orders"`
	} `json:"data"`
	Error interface{} `json:"error"`
}

type SubmitOrderResponse struct {
	Success bool `json:"success"`
	Data    struct {
		Order struct {
			ID         interface{} `json:"id"`
			StatusText string      `json:"status_text"`
			EachPrice  string      `json:"each_price"`
			Amount     string      `json:"amount"`
			Total      string      `json:"total"`
		} `json:"order"`
	} `json:"data"`
	Error interface{} `json:"error"`
}

func signParams(params map[string]string) string {
	var keys []string

	for k := range params {
		if k != "signature" {
			keys = append(keys, k)
		}
	}

	sort.Strings(keys)

	var parts []string
	for _, k := range keys {
		parts = append(parts, fmt.Sprintf("%s=%s", k, params[k]))
	}

	queryString := strings.Join(parts, "&")

	h := hmac.New(sha256.New, []byte(SECRET_KEY))
	h.Write([]byte(queryString))

	return hex.EncodeToString(h.Sum(nil))
}

func getBestAsk(symbolBase string, symbolQuote string) (string, error) {
	endpoint := BASE_URL + "/pro/capi/v1/markets/order-books"

	req, err := http.NewRequest("GET", endpoint, nil)
	if err != nil {
		return "", err
	}

	q := req.URL.Query()
	q.Add("base_coin", symbolBase)
	q.Add("quote_coin", symbolQuote)
	req.URL.RawQuery = q.Encode()

	req.Header.Set("Accept", "application/json")
	req.Header.Set("X-BIT24-APIKEY", API_KEY)

	client := &http.Client{}

	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("http error: %s", string(body))
	}

	var result OrderBookResponse

	err = json.Unmarshal(body, &result)
	if err != nil {
		return "", err
	}

	if !result.Success {
		return "", fmt.Errorf("order book error: %v", result.Error)
	}

	if len(result.Data.SellOrders) == 0 {
		return "", fmt.Errorf("no sell orders available")
	}

	return result.Data.SellOrders[0].Price, nil
}

func submitLimitBuyOrder(
	symbolBase string,
	symbolQuote string,
	amount string,
	price string,
) (*SubmitOrderResponse, error) {

	endpoint := BASE_URL + "/pro/capi/v1/orders/submit"

	params := map[string]string{
		"base_coin_symbol":  symbolBase,
		"quote_coin_symbol": symbolQuote,
		"type":              "1", // buy
		"category_type":     "0", // limit
		"price":             price,
		"amount":            amount,
	}

	params["signature"] = signParams(params)

	form := url.Values{}
	for k, v := range params {
		form.Add(k, v)
	}

	req, err := http.NewRequest(
		"POST",
		endpoint,
		strings.NewReader(form.Encode()),
	)

	if err != nil {
		return nil, err
	}

	req.Header.Set("Accept", "application/json")
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("X-BIT24-APIKEY", API_KEY)

	client := &http.Client{}

	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	if resp.StatusCode != http.StatusOK {
		fmt.Println("Server responded with error:")
		fmt.Println(string(body))
		return nil, fmt.Errorf("http error: %s", resp.Status)
	}

	var result SubmitOrderResponse

	err = json.Unmarshal(body, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

func buy2ADAOnIRT() error {
	base := "ADA"
	quote := "IRT"
	amount := "2"

	fmt.Println("Fetching best ask for ADA/IRT ...")

	bestAsk, err := getBestAsk(base, quote)
	if err != nil {
		return err
	}

	fmt.Printf("Best ask price (raw): %s IRT\n", bestAsk)

	fmt.Printf(
		"Placing limit buy order for %s ADA at %s IRT ...\n",
		amount,
		bestAsk,
	)

	response, err := submitLimitBuyOrder(
		base,
		quote,
		amount,
		bestAsk,
	)

	if err != nil {
		return err
	}

	if response.Success {
		order := response.Data.Order

		fmt.Println("Order placed successfully!")
		fmt.Printf("Order ID: %v\n", order.ID)
		fmt.Printf("Status: %s\n", order.StatusText)
		fmt.Printf("Price: %s IRT\n", order.EachPrice)
		fmt.Printf("Amount: %s ADA\n", order.Amount)
		fmt.Printf("Total: %s IRT\n", order.Total)

	} else {
		fmt.Printf("Order failed: %v\n", response.Error)
	}

	return nil
}

func main() {
	reader := bufio.NewReader(os.Stdin)

	fmt.Print("Please Enter Your API Key: ")
	apiKey, _ := reader.ReadString('\n')

	fmt.Print("Please Enter Your Secret Key: ")
	secretKey, _ := reader.ReadString('\n')

	API_KEY = strings.TrimSpace(apiKey)
	SECRET_KEY = strings.TrimSpace(secretKey)

	err := buy2ADAOnIRT()
	if err != nil {
		fmt.Printf("An error occurred: %v\n", err)
	}
}